protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArray.java

```diff
@@ -0,0 +1,380 @@
+   * been called without a subsequent `applyScoreChanges` call. This is because
+   * `onBlock` does not attempt to walk backwards through the tree and update the
+   * best child / best descendant links.
+   *
+   * @param justifiedRoot
+   * @return
+   */
+  public Bytes32 findHead(Bytes32 justifiedRoot) {
+    int justifiedIndex = indices.get(justifiedRoot);
+    ProtoNode justifiedNode = nodes.get(justifiedIndex);
+
+    int bestDescendantIndex = justifiedNode.getBestDescendantIndex().orElse(justifiedIndex);
+    ProtoNode bestNode = nodes.get(bestDescendantIndex);
+
+    // Perform a sanity check that the node is indeed valid to be the head.
+    if (!nodeIsViableForHead(bestNode)) {
+      throw new RuntimeException("ProtoArray: Best node is not viable for head");
+    }
+
+    return bestNode.getRoot();
+  }
+
+  /**
+   * Iterate backwards through the array, touching all nodes and their parents and potentially
+   * the bestChildIndex of each parent.
+   *
+   * The structure of the `nodes` array ensures that the child of each node is always
+   * touched before its parent.
+   *
+   * For each node, the following is done:
+   *
+   * - Update the node's weight with the corresponding delta.
+   * - Back-propagate each node's delta to its parents delta.
+   * - Compare the current node with the parents best child, updating it if the current node
+   * should become the best child.
+   * - If required, update the parents best descendant with the current node or its best descendant.
+   *
+   * @param deltas
+   * @param justifiedEpoch
+   * @param finalizedEpoch
+   */
+  public void applyScoreChanges(
+      List<Long> deltas, UnsignedLong justifiedEpoch, UnsignedLong finalizedEpoch) {
+    if (deltas.size() != indices.size()) {
+      throw new RuntimeException("ProtoArray: Invalid delta length");
+    }
+
+    if (!justifiedEpoch.equals(this.justifiedEpoch)
+        || !finalizedEpoch.equals(this.finalizedEpoch)) {
+      this.justifiedEpoch = justifiedEpoch;
+      this.finalizedEpoch = finalizedEpoch;
+    }
+
+    // Iterate backwards through all indices in `this.nodes`.
+    for (int nodeIndex = nodes.size() - 1; nodeIndex >= 0; nodeIndex--) {
+      ProtoNode node = nodes.get(nodeIndex);
+
+      // There is no need to adjust the balances or manage parent of the zero hash since it
+      // is an alias to the genesis block. The weight applied to the genesis block is
+      // irrelevant as we _always_ choose it and it's impossible for it to have a parent.
+      if (node.getRoot().equals(Bytes32.ZERO)) {
+        continue;
+      }
+
+      long nodeDelta = deltas.get(nodeIndex);
+      node.adjustWeight(nodeDelta);
+
+      if (node.getParentIndex().isEmpty()) {
+        continue;
+      }
+
+      int parentIndex = node.getParentIndex().get();
+      deltas.set(parentIndex, deltas.get(parentIndex) + nodeDelta);
+      maybeUpdateBestChildAndDescendant(parentIndex, nodeIndex);
+    }
```

Comment:
Took me a while to connect these two parts - I know we normally exit early, but I'd be very tempted to write this one as `if(node.getParentIndex().isPresent())`.  Meredith may disagree. :)

Ideally I'd use `node.getParentIndex().ifPresent(..)` but it needs access to the non-final `nodeIndex` which is slightly annoying so the if structure is probably the best approach.